### PR TITLE
3807 provide useful username constraint

### LINKED ISSUE #3807: Provide useful message about username constraints
## Describe the bug
The registration endpoint only functions correctly if the username consists of alphanumeric characters. If certain other characters, such as a "-" are included in the username, the endpoint will raise an exception, causing the registration page to hang indefinitely. The Island raise will the following error: 
```
Traceback (most recent call last):
  File "/home/msalvatore/tmp/.mount_InfecttiHGZb/opt/python3.11/lib/python3.11/site-packages/flask/app.py", line 1823, in full_dispatch_request
    rv = self.dispatch_request()
         ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/msalvatore/tmp/.mount_InfecttiHGZb/opt/python3.11/lib/python3.11/site-packages/flask/app.py", line 1799, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/msalvatore/tmp/.mount_InfecttiHGZb/opt/python3.11/lib/python3.11/site-packages/flask_restful/__init__.py", line 489, in wrapper
    resp = resource(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/msalvatore/tmp/.mount_InfecttiHGZb/opt/python3.11/lib/python3.11/site-packages/flask/views.py", line 107, in view
    return current_app.ensure_sync(self.dispatch_request)(**kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/msalvatore/tmp/.mount_InfecttiHGZb/opt/python3.11/lib/python3.11/site-packages/flask_restful/__init__.py", line 604, in dispatch_request
    resp = meth(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/msalvatore/tmp/.mount_InfecttiHGZb/usr/src/monkey_island/cc/services/authentication_service/flask_resources/utils.py", line 41, in decorated_function
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/msalvatore/tmp/.mount_InfecttiHGZb/usr/src/monkey_island/cc/services/authentication_service/flask_resources/register.py", line 43, in post
    response: ResponseValue = register()
                              ^^^^^^^^^^
  File "/home/msalvatore/tmp/.mount_InfecttiHGZb/opt/python3.11/lib/python3.11/site-packages/flask_security/decorators.py", line 622, in wrapper
    return f(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^
  File "/home/msalvatore/tmp/.mount_InfecttiHGZb/opt/python3.11/lib/python3.11/site-packages/flask_security/views.py", line 294, in register
    if form.validate_on_submit():
       ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/msalvatore/tmp/.mount_InfecttiHGZb/opt/python3.11/lib/python3.11/site-packages/flask_wtf/form.py", line 86, in validate_on_submit
    return self.is_submitted() and self.validate(extra_validators=extra_validators)
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/msalvatore/tmp/.mount_InfecttiHGZb/opt/python3.11/lib/python3.11/site-packages/flask_security/forms.py", line 612, in validate
    if not super().validate(**kwargs):
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/msalvatore/tmp/.mount_InfecttiHGZb/opt/python3.11/lib/python3.11/site-packages/wtforms/form.py", line 329, in validate
    return super().validate(extra)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/msalvatore/tmp/.mount_InfecttiHGZb/opt/python3.11/lib/python3.11/site-packages/wtforms/form.py", line 146, in validate
    if not field.validate(self, extra):
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/msalvatore/tmp/.mount_InfecttiHGZb/opt/python3.11/lib/python3.11/site-packages/wtforms/fields/core.py", line 242, in validate
    stop_validation = self._run_validation_chain(form, chain)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/msalvatore/tmp/.mount_InfecttiHGZb/opt/python3.11/lib/python3.11/site-packages/wtforms/fields/core.py", line 262, in _run_validation_chain
    validator(form, self)
  File "/home/msalvatore/tmp/.mount_InfecttiHGZb/opt/python3.11/lib/python3.11/site-packages/flask_security/forms.py", line 212, in unique_username
    form.existing_username_user = _datastore.find_user(
                                  ^^^^^^^^^^^^^^^^^^^^^
  File "/home/msalvatore/tmp/.mount_InfecttiHGZb/opt/python3.11/lib/python3.11/site-packages/flask_security/datastore.py", line 898, in find_user
    obj = self.user_model.objects(**query).first()
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/msalvatore/tmp/.mount_InfecttiHGZb/opt/python3.11/lib/python3.11/site-packages/mongoengine/queryset/base.py", line 296, in first
    result = queryset[0]
             ~~~~~~~~^^^
  File "/home/msalvatore/tmp/.mount_InfecttiHGZb/opt/python3.11/lib/python3.11/site-packages/mongoengine/queryset/base.py", line 203, in __getitem__
    queryset._cursor[key],
    ^^^^^^^^^^^^^^^^
  File "/home/msalvatore/tmp/.mount_InfecttiHGZb/opt/python3.11/lib/python3.11/site-packages/mongoengine/queryset/base.py", line 1681, in _cursor
    self._cursor_obj = self._collection.find(self._query, **self._cursor_args)
                                             ^^^^^^^^^^^
  File "/home/msalvatore/tmp/.mount_InfecttiHGZb/opt/python3.11/lib/python3.11/site-packages/mongoengine/queryset/base.py", line 1730, in _query
    self._mongo_query = self._query_obj.to_query(self._document)
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/msalvatore/tmp/.mount_InfecttiHGZb/opt/python3.11/lib/python3.11/site-packages/mongoengine/queryset/visitor.py", line 91, in to_query
    query = query.accept(QueryCompilerVisitor(document))
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/msalvatore/tmp/.mount_InfecttiHGZb/opt/python3.11/lib/python3.11/site-packages/mongoengine/queryset/visitor.py", line 184, in accept
    return visitor.visit_query(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/msalvatore/tmp/.mount_InfecttiHGZb/opt/python3.11/lib/python3.11/site-packages/mongoengine/queryset/visitor.py", line 80, in visit_query
    return transform.query(self.document, **query.query)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/msalvatore/tmp/.mount_InfecttiHGZb/opt/python3.11/lib/python3.11/site-packages/mongoengine/queryset/transform.py", line 124, in query
    value = field.prepare_query_value(op, value)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/msalvatore/tmp/.mount_InfecttiHGZb/opt/python3.11/lib/python3.11/site-packages/mongoengine/fields.py", line 171, in prepare_query_value
    value = re.escape(value)
            ^^^^^^^^^^^^^^^^
  File "/home/msalvatore/tmp/.mount_InfecttiHGZb/opt/python3.11/lib/python3.11/re/__init__.py", line 260, in escape
    pattern = str(pattern, 'latin1')
              ^^^^^^^^^^^^^^^^^^^^^^
TypeError: decoding to str: need a bytes-like object, NoneType found
```

This is two bugs in 1.

1. If "-" characters are invalid, the UI should inform the user before they attempt to submit the credentials for registration.
2. The registration page does not surface the 500 error to the user.

## To Reproduce
1. Start a fresh instance of the Island
1. Attempt to register a user with a hyphen in their name, such as "my-user"
1. Notice that the registration page hangs
1. Notice the traceback in the Island's log.

## Expected behavior
1. The name field should turn red and provide a useful hint about what characters are illegal if any characters other than alphanumerics are entered into the field.
2. The registration page should display a 500 error to the user if the Island returns one.

## Note

Note that both the registration page (frontend) and the registration API endpoint (backend) should validate that the username contains only alphanumerics and provide the caller with a meaningful error message. The backend cannot assume that the frontend has done proper validation, or that the request was issues by the frontend.

### PR DESCRIPTION
**Pull request by ilija-lazoroski**
# What does this PR do?

Fixes #3807.

Add any further explanations here.

## PR Checklist
* [x] Have you added an explanation of what your changes do and why you'd like to include them?
* [ ] Is the TravisCI build passing?
* [ ] Was the CHANGELOG.md updated to reflect the changes?
* [x] Was the documentation framework updated to reflect the changes?
* [x] Have you checked that you haven't introduced any duplicate code?

## Testing Checklist

* [x] Added relevant unit tests?
* [x] Do all unit tests pass?
* [ ] Do all end-to-end tests pass?
* [x] Any other testing performed?
    > Tested by {Running the Monkey locally with relevant config/running Island/...}
* [x] If applicable, add screenshots or log transcripts of the feature working

image
image

### CODE REVIEW

File: monkey/monkey_island/cc/services/authentication_service/flask_resources/register.py [hunk 1]
```diff
@@ -54,3 +59,18 @@ def post(self):
...
+    def _validate_username_and_password(self, username: str, password: str) -> List[str]:
+        validation_messages = []
+        security = current_app.extensions["security"]
+
+        username_util = security._username_util
+        password_util = security._password_util
+        username_validation = username_util.validate(username)[0]
+        password_validation = password_util.validate(password, False)[0]
+        if username_validation is not None or password_validation is not None:
+            validation_messages.append(username_validation)
```
Comment (mssalvatore) [inline monkey/monkey_island/cc/services/authentication_service/flask_resources/register.py:72]: `username_validation` might be `None`

File: monkey/monkey_island/cc/services/authentication_service/flask_resources/register.py [hunk 2]
```diff
@@ -40,6 +41,10 @@ def post(self):
         except Exception:
             return responses.make_response_to_invalid_request()
 
+        validation_messages = self._validate_username_and_password(username, password)
+        if validation_messages:
+            return responses.make_response_to_invalid_request(validation_messages)
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/services/authentication_service/flask_resources/register.py:46]:
```suggestion
        errors = self._validate_username_and_password(username, password)
        if errors:
            return responses.make_response_to_invalid_request(errors)
```

**Comment by VakarisZ** [comment]
This uses the same validators as flask_security, which is a good thing

**Comment by VakarisZ** [comment]
I don't get any error if I submit empty username

**Comment by ilija-lazoroski** [comment]
> I don't get any error if I submit empty username

You can't submit an empty username, it will just get you back to the username input field, which I think it is an UI bug.
You can confirm above, that if you hit `Let's go` no call to register endpoint will be done.

File: monkey/monkey_island/cc/services/authentication_service/flask_resources/register.py [hunk 3]
```diff
@@ -54,3 +59,19 @@ def post(self):
...
+
+    def _validate_username_and_password(self, username: str, password: str) -> List[str]:
+        validation_messages = []
+        security = current_app.extensions["security"]
+        username_util = security._username_util
+        password_util = security._password_util
+
+        username_validation = username_util.validate(username)[0]
+        password_validation = password_util.validate(password, False)[0]
+        if username_validation is not None:
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/services/authentication_service/flask_resources/register.py:71]:
```suggestion
        if username_errors

File: monkey/monkey_island/cc/services/authentication_service/flask_resources/register.py [hunk 4]
```diff
@@ -54,3 +59,19 @@ def post(self):
...
         return make_response(response)
+
+    def _validate_username_and_password(self, username: str, password: str) -> List[str]:
+        validation_messages = []
+        security = current_app.extensions["security"]
+        username_util = security._username_util
+        password_util = security._password_util
+
+        username_validation = username_util.validate(username)[0]
+        password_validation = password_util.validate(password, False)[0]
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/services/authentication_service/flask_resources/register.py:70]:
```suggestion
        username_errors = username_util.validate(username)[0]
        password_errors = password_util.validate(password, False)[0]
```

File: monkey/monkey_island/cc/services/authentication_service/flask_resources/register.py [hunk 5]
```diff
@@ -54,3 +59,19 @@ def post(self):
...
+        if username_validation is not None:
+            validation_messages.append(username_validation)
+
+        if password_validation is not None:
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/services/authentication_service/flask_resources/register.py:74]:
```suggestion
        if password_errors:
```

File: monkey/monkey_island/cc/services/authentication_service/flask_resources/register.py [hunk 6]
```diff
@@ -40,6 +41,10 @@ def post(self):
         except Exception:
             return responses.make_response_to_invalid_request()
 
+        errors = self._validate_username_and_password(username, password)
+        if errors:
+            return responses.make_response_to_invalid_request(errors)
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/services/authentication_service/flask_resources/register.py:46]: Request is valid, either password or username is not. `responses.make_response_to_invalid_request` is meant for invalid requests

Comment (ilija-lazoroski) [reply monkey/monkey_island/cc/services/authentication_service/flask_resources/register.py:46]: Flask is using 400 for the same bit. If we want to stay consistent we will go with that. Also check out this [link](https://stackoverflow.com/questions/6123425/rest-response-code-for-invalid-data). 

The only issue is the response which we get when we have invalid request and flask response.

One is `{"message": []}` other is `{"response": "errors": []}`

**Review by VakarisZ — APPROVED** [review]

**Pull request merged**